flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointBriefCalculator.java

```diff
@@ -0,0 +1,492 @@
+                    tasksToAck.put(vertex.getCurrentExecutionAttempt().getAttemptId(), vertex);
+
+                    if (someTasksMustBeTriggered) {
+                        boolean hasRunningPrecedentTasks =
+                                hasRunningPrecedentTasks(
+                                        runningTasksByVertex, prevJobEdges, vertex);
+
+                        if (!hasRunningPrecedentTasks) {
+                            tasksToTrigger.add(vertex.getCurrentExecutionAttempt());
+                        }
+                    }
+                } else {
+                    finishedTasks.add(vertex.getCurrentExecutionAttempt());
+                }
+            }
+        }
+
+        return new CheckpointBrief(
+                Collections.unmodifiableList(tasksToTrigger),
+                tasksToAck,
+                Collections.unmodifiableList(
+                        tasksToAck.size() == allTasks.size()
+                                ? allTasks
+                                : new ArrayList<>(tasksToAck.values())),
+                Collections.unmodifiableList(finishedTasks),
+                Collections.unmodifiableList(fullyFinishedJobVertex));
+    }
+
+    private boolean hasRunningPrecedentTasks(
+            Map<JobVertexID, JobVertexTaskSet> runningTasksByVertex,
+            List<JobEdge> prevJobEdges,
+            ExecutionVertex vertex) {
+        return IntStream.range(0, prevJobEdges.size())
+                .filter(
+                        i ->
+                                prevJobEdges.get(i).getDistributionPattern()
+                                        == DistributionPattern.POINTWISE)
+                .boxed()
+                .flatMap(i -> getPrecedentTasks(vertex, i).stream())
+                .anyMatch(
+                        precedentTask ->
+                                runningTasksByVertex
+                                        .get(precedentTask.getJobvertexId())
+                                        .contains(precedentTask.getID()));
+    }
+
+    private boolean someTasksMustBeTriggered(
+            Map<JobVertexID, JobVertexTaskSet> runningTasksByVertex,
+            JobVertexTaskSet runningTasks,
+            List<JobEdge> prevJobEdges) {
+        if (runningTasks.containsNoTasks()) {
+            // if this task doesn't have any running tasks, we cannot trigger anything
+            return false;
+        }
+        for (JobEdge jobEdge : prevJobEdges) {
+            DistributionPattern distributionPattern = jobEdge.getDistributionPattern();
+            JobVertexTaskSet sourceRunningTasks =
+                    runningTasksByVertex.get(jobEdge.getSource().getProducer().getID());
+
+            if ((distributionPattern == DistributionPattern.ALL_TO_ALL
+                            && !sourceRunningTasks.containsNoTasks())
+                    || (distributionPattern == DistributionPattern.POINTWISE
+                            && sourceRunningTasks.containsAllTasks())) {
```

Comment:
Could you extract a function
```
private boolean hasConnection(DistributionPattern distribution, JobVertexTaskSet runningTasks) {
    switch (distribution) {
        case ALL_TO_ALL:
            return !runningTasks.containsAllTasks();
        case POINTWISE:
            return runningTasks.containsNoTasks();
        default:
            throw new IllegalArgumentException(distribution.name());
    }
}
```
and use it here and in `isFinishedAccordingToDescendants`?

Reply:
I'm a bit concerns here in that the condition used for calculate trigger tasks and calculate finished tasks are a bit different:
1. For calculating tasks to trigger, it would check if it is ALL_TO_ALL connected to some running tasks or POINTWISE connected to all running tasks.
2. For judge if a task is finished, it would check if it is ALL_TO_ALL connected to some finished tasks (namely not all tasks are running) or POINTWISE connected to all finished tasks.

Thus it seems to be not easy to extract a common method~?

Reply:
Yes, you are right.
I'd still extract these checks into a method or two methods:
```
private boolean haveConnection(
        DistributionPattern distribution, JobVertexTaskSet tasks, boolean active) {
    switch (distribution) {
        case POINTWISE:
            return active ? tasks.containsAllTasks() : tasks.containsNoTasks();
        case ALL_TO_ALL:
            return active ? !tasks.containsNoTasks() : !tasks.containsAllTasks();
        default:
            throw new IllegalArgumentException(distribution.name());
    }
}
```

or

```
private boolean hasActiveUpstream(DistributionPattern distribution, JobVertexTaskSet up) {
    return (distribution == ALL_TO_ALL && !up.containsNoTasks())
            || (distribution == POINTWISE && up.containsAllTasks());
}

private boolean hasInactiveDownstream(DistributionPattern distribution, JobVertexTaskSet down) {
    return (distribution == ALL_TO_ALL && !down.containsAllTasks())
            || (distribution == POINTWISE && down.containsNoTasks());
}
```
WDYT?

Reply:
Do we really need to differentiate between ALL_TO_ALL and POINTWISE? Why can't we use the same branch for POINTWISE as we use for  ALL_TO_ALL?

Reply:
This is due to that for example, when we are judge whether a task is in fact finished:
1. If it is connected with downstream tasks with ALL_TO_ALL edges, then all the tasks of this vertex would need to send EndOfPartition to all its descendants, thus if _any_ of the descendants is finished, we know all the tasks of this vertex has sent EndOfPartition (otherwise the downstream tasks cannot be finished).
2. But it would not hold for pointwise edges since each task of this vertex only connects to a part of the downstream tasks. Only when _all_ the tasks of the downstream tasks are finished, we could judge that all the tasks of this vertex has finished. 

These two would serve as a fast path so that we do not need to judge each task separately if possible. But if it cannot decide, we would then head to judge each task separately~

Reply:
I understand, thanks!